File: internal/usage/codex.go
package usage

import (
	"bufio"
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/ashon/ax/internal/agent"
)

// codexAgentName is the synthetic agent label attached to every Codex-derived
// transcriptSeries so per-workspace rollups can still report "codex" activity
// even without a sub-agent identifier like Claude's agentId.
const codexAgentName = "codex"

// codexRecord is the on-disk shape for one JSONL line in a Codex session
// rollout file. Codex emits both session metadata (`type=session_meta`) and
// per-event messages (`type=event_msg`); we only care about the meta header
// and `token_count` events within that stream.
type codexRecord struct {
	Timestamp time.Time       `json:"timestamp"`
	Type      string          `json:"type"`
	Payload   json.RawMessage `json:"payload"`
}

type codexSessionMetaPayload struct {
	ID        string    `json:"id"`
	Timestamp time.Time `json:"timestamp"`
	Cwd       string    `json:"cwd"`
}

type codexEventPayload struct {
	Type string          `json:"type"`
	Info *codexTokenInfo `json:"info"`
}

type codexTokenInfo struct {
	TotalTokenUsage codexTokenUsage `json:"total_token_usage"`
	LastTokenUsage  codexTokenUsage `json:"last_token_usage"`
	// ModelContextWindow is reported but we do not currently persist it.
}

// codexTokenUsage mirrors the `total_token_usage` / `last_token_usage` shape
// Codex writes. Note that `input_tokens` is INCLUSIVE of `cached_input_tokens`
// (confirmed via total_tokens == input_tokens + output_tokens); we subtract
// `cached_input_tokens` before mapping into the runtime-neutral `Tokens`
// struct so cumulative totals stay consistent across runtimes.
type codexTokenUsage struct {
	InputTokens         int64 `json:"input_tokens"`
	CachedInputTokens   int64 `json:"cached_input_tokens"`
	OutputTokens        int64 `json:"output_tokens"`
	ReasoningOutputTok  int64 `json:"reasoning_output_tokens"`
	TotalTokens         int64 `json:"total_tokens"`
}

func (u codexTokenUsage) toTokens() Tokens {
	input := u.InputTokens - u.CachedInputTokens
	if input < 0 {
		input = 0
	}
	return Tokens{
		Input:     input,
		Output:    u.OutputTokens,
		CacheRead: u.CachedInputTokens,
		// Codex has no distinct "cache creation" concept; leave at zero.
	}
}

// discoverCodexSessions returns every rollout JSONL under a workspace's
// Codex home whose mtime falls inside [since, until]. Codex organises files
// as $CODEX_HOME/sessions/YYYY/MM/DD/rollout-*.jsonl; we walk the full tree
// rather than trying to predict dates, since session files can be written
// across day boundaries.
func discoverCodexSessions(sessionsDir string, since, until time.Time) ([]string, error) {
	paths := make([]string, 0, 16)
	err := filepath.WalkDir(sessionsDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			if errors.Is(err, os.ErrNotExist) {
				return filepath.SkipAll
			}
			return err
		}
		if d.IsDir() {
			return nil
		}
		if filepath.Ext(path) != ".jsonl" {
			return nil
		}
		info, statErr := d.Info()
		if statErr != nil {
			return nil
		}
		if !since.IsZero() && info.ModTime().Before(since) {
			return nil
		}
		paths = append(paths, path)
		return nil
	})
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	sort.Strings(paths)
	return paths, nil
}

// scanCodexTranscript reads a single Codex session file and produces a
// transcriptSeries with per-turn buckets. Turns here correspond to
// token_count events whose `info.last_token_usage` is populated; Codex also
// emits a no-op token_count before any model call that only carries a
// rate-limit snapshot, which we skip.
func scanCodexTranscript(path string, q HistoryQuery) (*transcriptSeries, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 4*1024*1024)
	agg := NewAggregator()
	buckets := map[time.Time]*Bucket{}
	series := &transcriptSeries{
		agent:      codexAgentName,
		transcript: path,
	}

	for scanner.Scan() {
		var rec codexRecord
		if err := json.Unmarshal(scanner.Bytes(), &rec); err != nil {
			continue
		}
		switch rec.Type {
		case "session_meta":
			var meta codexSessionMetaPayload
			if err := json.Unmarshal(rec.Payload, &meta); err != nil {
				continue
			}
			if series.sessionID == "" {
				series.sessionID = meta.ID
			}
			if series.cwd == "" && meta.Cwd != "" {
				series.cwd = cleanPath(meta.Cwd)
			}
		case "event_msg":
			var ev codexEventPayload
			if err := json.Unmarshal(rec.Payload, &ev); err != nil {
				continue
			}
			if ev.Type != "token_count" || ev.Info == nil {
				continue
			}
			delta := ev.Info.LastTokenUsage.toTokens()
			if delta == (Tokens{}) {
				continue
			}
			parsed := parsedRecord{
				SessionID: series.sessionID,
				Cwd:       series.cwd,
				Timestamp: rec.Timestamp,
				Model:     codexAgentName,
				Tokens:    delta,
				HasUsage:  true,
			}
			effect := agg.Ingest(parsed)
			if rec.Timestamp.IsZero() || rec.Timestamp.Before(q.Since) || !rec.Timestamp.Before(q.Until) {
				continue
			}
			if effect.TurnDelta == 0 && effect.TokensDelta == (Tokens{}) {
				continue
			}
			start := rec.Timestamp.UTC().Truncate(q.BucketSize)
			b := buckets[start]
			if b == nil {
				b = &Bucket{Start: start, End: start.Add(q.BucketSize)}
				buckets[start] = b
			}
			if effect.TokensDelta != (Tokens{}) {
				b.Tokens = b.Tokens.Add(effect.TokensDelta)
				b.Total += effect.TokensDelta.Total()
			}
			if effect.TurnDelta != 0 {
				b.Turns += effect.TurnDelta
			}
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	snap := agg.Snapshot("", path)
	if !snap.Available && series.sessionID == "" {
		return nil, nil
	}
	series.current = CurrentSnapshot{
		LastActivity:       snap.LastActivity,
		CurrentContext:     snap.CurrentContext,
		CurrentTotal:       snap.CurrentContext.Total(),
		CurrentMCPProxy:    snap.CurrentMCP,
		CurrentModel:       snap.CurrentModel,
		CumulativeTotals:   snap.CumulativeTotals,
		CumulativeTotal:    snap.CumulativeTotals.Total(),
		CumulativeMCPProxy: snap.CumulativeMCP,
		Turns:              snap.Turns,
	}
	series.buckets = sortBuckets(buckets)
	return series, nil
}

// codexScanResult captures the per-workspace scan outcome for Codex. Unlike
// the Claude code path, each session file belongs to a known workspace
// (Codex runs with a workspace-specific CODEX_HOME), so attribution is
// deterministic and no hint matching is required.
type codexScanResult struct {
	homeExists    bool
	sessionsFound bool
	series        []*transcriptSeries
}

// scanCodexForBinding returns every Codex transcript series attributable to
// one ax workspace. Missing CODEX_HOME directories yield an empty result
// rather than an error so Claude-only workspaces still work.
func scanCodexForBinding(binding WorkspaceBinding, q HistoryQuery) (codexScanResult, error) {
	var res codexScanResult
	if binding.Name == "" || binding.Dir == "" {
		return res, nil
	}
	home, err := agent.CodexHomePath(binding.Name, binding.Dir)
	if err != nil {
		return res, err
	}
	if _, err := os.Stat(home); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return res, nil
		}
		return res, err
	}
	res.homeExists = true

	sessionsDir := filepath.Join(home, "sessions")
	paths, err := discoverCodexSessions(sessionsDir, q.Since, q.Until)
	if err != nil {
		return res, err
	}
	if len(paths) == 0 {
		return res, nil
	}
	res.sessionsFound = true
	for _, path := range paths {
		s, err := scanCodexTranscript(path, q)
		if err != nil {
			return res, err
		}
		if s != nil {
			res.series = append(res.series, s)
		}
	}
	return res, nil
}

File: crates/ax-usage/src/codex.rs
//! Codex session rollout JSONL parser.
//!
//! Codex emits two line shapes we care about:
//!   - `type=session_meta` once at the top of the file, carrying the
//!     session id, cwd, and timestamp.
//!   - `type=event_msg` repeatedly; we only attribute `token_count`
//!     events whose `info` is populated. Codex also emits `token_count`
//!     with `info: null` before any model call (rate-limit pulse) —
//!     those are skipped.
//!
//! A populated `token_count` event carries both `total_token_usage`
//! (cumulative since session start) and `last_token_usage` (per-turn
//! delta); we treat the latter as the turn's tokens because the
//! aggregator expects per-turn deltas, not cumulative snapshots.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::Deserialize;

use ax_proto::usage::{MCPProxyMetrics, Tokens};

use crate::aggregator::Aggregator;
use crate::history::{Bucket, CurrentSnapshot, HistoryError, HistoryQuery, TranscriptSeries};
use crate::parse::ParsedRecord;

/// Synthetic agent label attached to every Codex-derived parsed record.
pub const CODEX_AGENT_NAME: &str = "codex";

/// Classification of a single JSONL line from a Codex rollout.
#[derive(Debug, Clone)]
pub enum CodexLine {
    /// First line of the file: session identity.
    SessionMeta {
        session_id: String,
        cwd: String,
        timestamp: Option<DateTime<Utc>>,
    },
    /// A populated `token_count` event (rate-limit-only pulses are skipped).
    TokenCount {
        timestamp: Option<DateTime<Utc>>,
        /// Cumulative usage across the session at this moment. Not used
        /// by the aggregator but exposed for callers that want to
        /// display context-window headroom.
        cumulative: Tokens,
        /// Per-turn delta that the aggregator should ingest.
        delta: Tokens,
    },
    /// Any other line we don't attribute (`session_footer`, tool events,
    /// text deltas, rate-limit-only pulses).
    Other,
}

#[derive(Debug, thiserror::Error)]
pub enum CodexParseError {
    #[error("decode codex line: {0}")]
    Json(#[from] serde_json::Error),
}

/// Parse one JSONL line from a Codex session file.
pub fn parse_codex_line(data: &[u8]) -> Result<CodexLine, CodexParseError> {
    let raw: RawCodexRecord = serde_json::from_slice(data)?;
    Ok(classify(raw))
}

/// Build the synthetic [`ParsedRecord`] that the [`crate::Aggregator`]
/// ingests for a Codex `token_count` event. `session_id` / `cwd` come
/// from an earlier `SessionMeta` line on the same file.
#[must_use]
pub fn parsed_record_from_codex(
    session_id: &str,
    cwd: &str,
    timestamp: Option<DateTime<Utc>>,
    delta: Tokens,
) -> ParsedRecord {
    ParsedRecord {
        session_id: session_id.to_owned(),
        cwd: cwd.to_owned(),
        timestamp,
        model: CODEX_AGENT_NAME.to_owned(),
        tokens: delta,
        has_usage: true,
        ..ParsedRecord::default()
    }
}

// ---------- serde view ----------

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawCodexRecord {
    timestamp: Option<DateTime<Utc>>,
    #[serde(rename = "type")]
    line_type: String,
    payload: serde_json::Value,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawSessionMetaPayload {
    id: String,
    cwd: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawEventPayload {
    #[serde(rename = "type")]
    event_type: String,
    info: Option<RawTokenInfo>,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawTokenInfo {
    total_token_usage: RawCodexUsage,
    last_token_usage: RawCodexUsage,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
#[allow(clippy::struct_field_names)] // JSON keys verbatim from codex
struct RawCodexUsage {
    input_tokens: i64,
    cached_input_tokens: i64,
    output_tokens: i64,
    #[allow(dead_code)]
    reasoning_output_tokens: i64,
    #[allow(dead_code)]
    total_tokens: i64,
}

impl RawCodexUsage {
    fn to_tokens(&self) -> Tokens {
        let input = (self.input_tokens - self.cached_input_tokens).max(0);
        Tokens {
            input,
            output: self.output_tokens,
            cache_read: self.cached_input_tokens,
            cache_creation: 0,
        }
    }
}

fn classify(raw: RawCodexRecord) -> CodexLine {
    match raw.line_type.as_str() {
        "session_meta" => {
            let meta: RawSessionMetaPayload =
                serde_json::from_value(raw.payload).unwrap_or_default();
            CodexLine::SessionMeta {
                session_id: meta.id,
                cwd: meta.cwd,
                timestamp: raw.timestamp,
            }
        }
        "event_msg" => {
            let ev: RawEventPayload = serde_json::from_value(raw.payload).unwrap_or_default();
            if ev.event_type != "token_count" {
                return CodexLine::Other;
            }
            let Some(info) = ev.info else {
                return CodexLine::Other;
            };
            let delta = info.last_token_usage.to_tokens();
            if delta == Tokens::default() {
                return CodexLine::Other;
            }
            CodexLine::TokenCount {
                timestamp: raw.timestamp,
                cumulative: info.total_token_usage.to_tokens(),
                delta,
            }
        }
        _ => CodexLine::Other,
    }
}

// ---------- session discovery + scan ----------

/// Outcome of a Codex scan for one workspace binding. The status flags
/// let the caller pick the right `unavailable_reason` when both Claude
/// and Codex come up empty.
#[derive(Debug, Default)]
pub(crate) struct CodexScanResult {
    pub home_exists: bool,
    pub sessions_found: bool,
    pub series: Vec<TranscriptSeries>,
}

/// Walk `sessions_dir` for every `rollout-*.jsonl` whose modification
/// time falls at or after `since`. Returns an empty list when the
/// directory doesn't exist. Codex organises files under
/// `YYYY/MM/DD/rollout-*.jsonl`; we walk the full tree rather than
/// predicting dates so cross-midnight sessions get picked up.
pub(crate) fn discover_codex_sessions(
    sessions_dir: &Path,
    since: DateTime<Utc>,
) -> Result<Vec<PathBuf>, HistoryError> {
    if !sessions_dir.exists() {
        return Ok(Vec::new());
    }
    let mut out = Vec::new();
    walk_sessions(sessions_dir, since, &mut out)?;
    out.sort();
    Ok(out)
}

fn walk_sessions(
    dir: &Path,
    since: DateTime<Utc>,
    out: &mut Vec<PathBuf>,
) -> Result<(), HistoryError> {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(e) => {
            return Err(HistoryError::Read {
                path: dir.display().to_string(),
                source: e,
            })
        }
    };
    for entry in entries {
        let entry = entry.map_err(|e| HistoryError::Read {
            path: dir.display().to_string(),
            source: e,
        })?;
        let path = entry.path();
        let file_type = entry.file_type().map_err(|e| HistoryError::Read {
            path: path.display().to_string(),
            source: e,
        })?;
        if file_type.is_dir() {
            walk_sessions(&path, since, out)?;
            continue;
        }
        if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
            continue;
        }
        if since.timestamp() > 0 {
            let Ok(meta) = std::fs::metadata(&path) else {
                continue;
            };
            let Ok(modified) = meta.modified() else {
                continue;
            };
            let modified: DateTime<Utc> = modified.into();
            if modified < since {
                continue;
            }
        }
        out.push(path);
    }
    Ok(())
}

/// Scan every Codex session under `home_dir` for `workspace`. Missing
/// `home_dir` yields an empty result (homes only exist once codex has
/// run against that workspace at least once).
pub(crate) fn scan_codex_for_binding(
    workspace: &str,
    home_dir: &Path,
    query: &HistoryQuery,
) -> Result<CodexScanResult, HistoryError> {
    let _ = workspace; // reserved for future multi-agent labeling
    let mut res = CodexScanResult::default();
    if !home_dir.exists() {
        return Ok(res);
    }
    res.home_exists = true;
    let sessions_dir = home_dir.join("sessions");
    let paths = discover_codex_sessions(&sessions_dir, query.since)?;
    if paths.is_empty() {
        return Ok(res);
    }
    res.sessions_found = true;
    for path in paths {
        if let Some(series) = scan_codex_transcript(&path, query)? {
            res.series.push(series);
        }
    }
    Ok(res)
}

fn scan_codex_transcript(
    path: &Path,
    query: &HistoryQuery,
) -> Result<Option<TranscriptSeries>, HistoryError> {
    use std::io::BufRead;

    let file = std::fs::File::open(path).map_err(|e| HistoryError::Read {
        path: path.display().to_string(),
        source: e,
    })?;
    let reader = std::io::BufReader::new(file);

    let mut agg = Aggregator::new();
    let mut buckets: BTreeMap<DateTime<Utc>, Bucket> = BTreeMap::new();
    let mut series = TranscriptSeries {
        agent: CODEX_AGENT_NAME.to_owned(),
        transcript: path.to_path_buf(),
        ..TranscriptSeries::default()
    };

    for line in reader.lines() {
        let line = line.map_err(|e| HistoryError::Read {
            path: path.display().to_string(),
            source: e,
        })?;
        let Ok(parsed) = parse_codex_line(line.as_bytes()) else {
            continue;
        };
        match parsed {
            CodexLine::SessionMeta {
                session_id, cwd, ..
            } => {
                if series.session_id.is_empty() && !session_id.is_empty() {
                    series.session_id = session_id;
                }
                if series.cwd.is_empty() && !cwd.is_empty() {
                    series.cwd = cwd;
                }
            }
            CodexLine::TokenCount {
                timestamp, delta, ..
            } => {
                let rec =
                    parsed_record_from_codex(&series.session_id, &series.cwd, timestamp, delta);
                let effect = agg.ingest(&rec);
                let Some(ts) = timestamp else { continue };
                if ts < query.since || ts >= query.until {
                    continue;
                }
                if effect.turn_delta == 0 && effect.tokens_delta == Tokens::default() {
                    continue;
                }
                let start = truncate_to_bucket(ts, query.bucket_size);
                let bucket_duration = chrono::Duration::from_std(query.bucket_size)
                    .unwrap_or(chrono::Duration::zero());
                let b = buckets.entry(start).or_insert_with(|| Bucket {
                    start,
                    end: start + bucket_duration,
                    ..Bucket::default()
                });
                if effect.tokens_delta != Tokens::default() {
                    b.tokens = b.tokens + effect.tokens_delta;
                    b.total += effect.tokens_delta.total();
                }
                if effect.turn_delta != 0 {
                    b.turns += effect.turn_delta;
                }
            }
            CodexLine::Other => {}
        }
    }

    let snap = agg.snapshot("", &series.transcript.display().to_string());
    if !snap.available && series.session_id.is_empty() {
        return Ok(None);
    }
    series.current = CurrentSnapshot {
        last_activity: snap.last_activity,
        current_context: snap.current_context,
        current_total: snap.current_context.total(),
        current_mcp_proxy: snap.current_mcp,
        current_model: snap.current_model,
        cumulative_totals: snap.cumulative_totals,
        cumulative_total: snap.cumulative_totals.total(),
        cumulative_mcp_proxy: snap.cumulative_mcp,
        turns: snap.turns,
    };
    series.buckets = buckets.into_values().collect();
    Ok(Some(series))
}

fn truncate_to_bucket(ts: DateTime<Utc>, bucket: std::time::Duration) -> DateTime<Utc> {
    let secs = bucket.as_secs() as i64;
    if secs == 0 {
        return ts;
    }
    let epoch = ts.timestamp();
    let truncated = epoch - epoch.rem_euclid(secs);
    DateTime::<Utc>::from_timestamp(truncated, 0).unwrap_or(ts)
}

#[allow(dead_code)]
fn _use_mcp(_: MCPProxyMetrics) {}
